retry.go

```diff
@@ -287,4 +287,55 @@ func (t *Transaction[T]) Process(state T) (T, error) {
 	return state, err
 }
 
-// throttle ?
+type throttle struct {
+	mu         *sync.Mutex
+	timer      *time.Timer
+	needInvoke bool
+	interval   time.Duration
+	callbacks  []func()
+}
+
+func (th *throttle) throttledFunc() {
+	th.mu.Lock()
+	defer th.mu.Unlock()
+	th.needInvoke = true
+	if th.timer == nil {
+		th.timer = time.AfterFunc(th.interval, func() {
+			th.purge(false)
+		})
+	}
+}
+
+func (th *throttle) purge(forcePurge bool) {
+	th.mu.Lock()
+	defer th.mu.Unlock()
+
+	if th.timer != nil {
+		th.timer.Stop()
+	}
+
+	if th.needInvoke || forcePurge {
+		for _, f := range th.callbacks {
+			go f()
+		}
+		th.needInvoke = false
+		th.timer = time.AfterFunc(th.interval, func() {
+			th.purge(false)
+		})
+	} else {
+		th.timer = nil
+	}
+}
+
+// NewThrottle creates a throttled instance that invokes given functions only once in every interval.
+// This returns 2 functions, First one is throttled function and Second one is purge function which invokes given functions immediately and reset interval timer.
+func NewThrottle(interval time.Duration, f ...func()) (func(), func()) {
```

Comment:
I'm also think about a `NewThrottle(100ms, 3, cb)` that would trigger the first 3 calls, then ignore others during 100ms. Like we do for a basic rate limiter.

Maybe it is a different helper.

WDYT?

Reply:
> I'm also think about a `NewThrottle(100ms, 3, cb)` that would trigger the first 3 calls, then ignore others during 100ms. Like we do for a basic rate limiter.
> 
> Maybe it is a different helper.
> 
> WDYT?

That would be a interesting option, but I am also afraid that might makes the function's signature complicated.
So I think, it's better to be implemented in other helper or function with different signature.

Reply:
2 helpers `NewThrottle` and `NewThrottleWithCount` ?

Reply:
```NewThrottleWithCount``` looks nice,
I'll add it